### ISSUE TITLE
Debouncing updates

### ISSUE THREAD

**Issue by sw-yx** [paoloricciuti/sveltekit-search-params#10]
it can be nasty to the browser history when we type and the url bar is updated on every keystroke.

could we offer an easy way (probably not default, but recommended) to debounce the updates?

**Comment by paoloricciuti** [comment]
This could very well be an optional third parameter in the store creator.

What i'm thinking tho is that the url might still update with the replaceState option set to true and only after the debounce (which defaults to 0) we add a new history entry, we can even avoid to use sveltekit apis and just do a

```ts
history.pushState(history.state, null)
```

This is because i think is still important to update the url right away with every change but we can avoid filling the history by doing this. I think this would be the best in terms of UX.

The propsed api will look like this

```svelte
<script>
    import {queryParam, ssp} from "sveltekit-search-params";
    
    const test = queryParam("test", ssp.string(), {
      debounceHistory: 300, //with this the URL will update on every keystroke but a new history will be debounced to 300ms
   })
</script>

<input bind:value={$test} />
```

We could even add another option to completely disable pushing history, something like

```svelte
<script>
    import {queryParam, ssp} from "sveltekit-search-params";
    
    const test = queryParam("test", ssp.string(), {
       pushHistory: false, //no new history will be created for this query param
   })
</script>

<input bind:value={$test} />
```

The only thing i don't really like about this is that to pass the options you have to pass the encoder and decoder. I could put the options as an optional parameter in the second spot of the arguments but this would be a breaking change (it may still be worth given that we are still not at 1.0?).

Ideas on the API, the naming etc? I'll start to work on this tomorrow morning probably.

**Comment by sw-yx** [comment]
i think this is a good API yes but i wonder if there is a lower level api we can offer that is less declarative but more generically powerful (maybe also replaces encoder?)

```svelte
<script>
    import {queryParam, ssp} from "sveltekit-search-params";

    // debounce fn here

    const test = queryParam("test", ssp.string(), (state, saveState) => {
      debounce(() => saveState(state), 300)
   })
</script>

<input bind:value={$test} />
```

**Comment by paoloricciuti** [comment]
I like the idea of a more powerful api but I dislike that this is the only option...what if we provide both? We could provide a set of options as the third argument and provide listener for the update:

```svelte
<script>
    import {queryParam, ssp} from "sveltekit-search-params";
    
    const test = queryParam("test", ssp.string(), {
      debounceHistory: 300, //with this the URL will update on every keystroke but a new history will be debounced to 300ms
   })
   test.on("update", (state, saveState)=>{
        //Do your powerful stuff
   })
</script>

<input bind:value={$test} />
```

**Comment by sw-yx** [comment]
up to you :) if in doubt, just do the simple thing first

**Comment by paoloricciuti** [comment]
Is just that I think the history issue is something very real and devs shouldn't be forced to write always the same code for something like this.

The idea of a callback in the "lifetime" is tempting tho

**Comment by sw-yx** [comment]
yup yup. do whatever is simple to you, we dont need to overthink this, i was just offering the obvious alternative api

**Comment by paoloricciuti** [comment]
@sw-yx updated the package to 1.0.0 with the new options added. Let me know if it fits your needs.

I'll probably add the listener in another task.

**Comment by sw-yx** [comment]
oh nice! thanks!
